### PR TITLE
PMEM Storage Level Implementation

### PR DESCRIPTION
**Pull request by yeyuqiang**
- Use pre-installed oap-common package
- PMEM_ONLY Storage Level
- PMEM_AND_DISK Storage Level

Passing simple kmeans test

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 1]
```diff
@@ -86,6 +91,7 @@ private[spark] class UnifiedMemoryManager private[memory] (
       memoryMode: MemoryMode): Long = synchronized {
```
Comment (winningsix) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:91]: cc @offthewall123 @jikunshang W.R.T out of memory issue, my understanding is we cached large size fiberCache. Actually Spark only needs batch based access, saying 1024 rows of data. Could we introduce a memory shrink mechanism to release accessed "blocks"? So FiberCache will be multiple batches, each batch is mapping to a column vector in Spark. The very first step is to make allocate synchronized. We can always wait to have free memory ready since Spark is continuing consuming memory.

Ideas?

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 2]
```diff
@@ -169,6 +180,8 @@ private[spark] class UnifiedMemoryManager private[memory] (
       return false
     }
     if (numBytes > storagePool.memoryFree) {
+      // When in PMem mode, borrowing from execution memory or eviction will not happen
```
Comment (winningsix) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:183]: What's the reason we stop borrowing memory from ```execMemPool```?

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 3]
```diff
@@ -161,6 +167,11 @@ private[spark] class UnifiedMemoryManager private[memory] (
         offHeapExecutionMemoryPool,
```
Comment (winningsix) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:167]: Hmm, do we need to apply different strategy for unroll memory? That memory should requires RAM instead of PMem.

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 4]
```diff
@@ -86,6 +91,7 @@ private[spark] class UnifiedMemoryManager private[memory] (
       memoryMode: MemoryMode): Long = synchronized {
     assertInvariants()
     assert(numBytes >= 0)
+    require(memoryMode != MemoryMode.PMEM, "PMem can not used as execution memory")
```
[the remark this answers is not in the document]
Comment (yma11) [reply core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:94]: @winningsix "When borrowing memory from storage, it will first move to Optane PMem and then spill." do you mean RDD cache will move to Optane Pmem and then spill? In current RDD cache solution, DRAM storage execution is allocated even we doesn't use it since AEP pool replaced this pool in real. Execution memory still can borrow memory from storage dram, not AEP.  I mean, when check whether this borrow is allowed, aep pool size is used instead of dram storage pool size in logic memory level. but if the borrow is allowed, memory is allocated from dram but not AEP.

[the remark this answers is not in the document]
Comment (winningsix) [reply core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:94]: But it seems we stop this behavior (exec memory borrows from storage mem) in ```UnifiedMemoryManager``` L183. Let's discuss it today.

File: core/src/main/scala/org/apache/spark/storage/StorageLevel.scala [hunk 5]
```diff
@@ -181,6 +196,7 @@ object StorageLevel {
     case "MEMORY_AND_DISK_SER" => MEMORY_AND_DISK_SER
     case "MEMORY_AND_DISK_SER_2" => MEMORY_AND_DISK_SER_2
     case "OFF_HEAP" => OFF_HEAP
+    case "PMEM" => PMEM_AND_DISK
```
Comment (winningsix) [inline core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:199]: Should be PMEM_AND_DISK?

File: pom.xml [hunk 6]
```diff
@@ -293,6 +295,11 @@
   </dependencies>
   <dependencyManagement>
     <dependencies>
+      <dependency>
+        <groupId>com.intel</groupId>
+        <artifactId>oap-common</artifactId>
+        <version>${oap.common.version}</version>
```
Comment (winningsix) [inline pom.xml:301]: Let’s use OAP version directly.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/memory/PMemMemoryAllocator.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package org.apache.spark.unsafe.memory;
+
+import com.intel.oap.unsafe.PersistentMemoryPlatform;
```
Comment (winningsix) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/memory/PMemMemoryAllocator.java:3]: We should put those under OAP COMMON package. It may need some code refactoring. You can do it in another PR.

File: core/src/main/scala/org/apache/spark/memory/MemoryManager.scala [hunk 8]
```diff
@@ -48,6 +48,8 @@ private[spark] abstract class MemoryManager(
   @GuardedBy("this")
   protected val offHeapStorageMemoryPool = new StorageMemoryPool(this, MemoryMode.OFF_HEAP)
   @GuardedBy("this")
+  protected val pmemStorageMemoryPool = new StorageMemoryPool(this, MemoryMode.PMEM)
```
Comment (winningsix) [inline core/src/main/scala/org/apache/spark/memory/MemoryManager.scala:51]: THe same as above. Let's try to not touch MemoryManager.

**Comment by winningsix** [comment]
@yeyuqiang  when will it use DRAM storage memory in this PR? My understanding is they will only allocate PMem memory once Memory mode is set to PMem.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/memory/PMemMemoryAllocator.java (hunk 7 above)
Comment (winningsix) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/memory/PMemMemoryAllocator.java:3]: Yes. It should happen in OAP side.

**Comment by yeyuqiang** [comment]
> > > @yeyuqiang when will it use DRAM storage memory in this PR? My understanding is they will only allocate PMem memory once Memory mode is set to PMem.
> 
> > Yes, this PR is replace DRAM by PMem
> 
> It will be tricky about the expected behavior for storage memory.
> 
> Let me give an example here.
> 
> Total 100GB DRAM, 80GB for exec mem, 20GB for storage mem.
> 
> with Storage PMem extension, we have two options:
> 
> 1. 100GB will be total for DRAM, in that case it will be quite complex in configuration. Current code is at this case. But configuration side is confusing customer. To be honest, I didn't have a good idea here for an expected configuration behavior.
> 2. still the same for DRAM side and 1TB total for storage. This will be very clear to end users in terms of configuration. And it's our new design as last discussion.
> 
> So what's your ideas here for option 1?

Why we need to struggle with option1(current code) here since we want to move forward to option 2.

**Comment by winningsix** [comment]
> Why we need to struggle with option1(current code) here since we want to move forward to option 2.

No, we don't have to. But if we have some short cut ideas to make option 1 work. We can work on it. Otherwise, I am suggesting to refactor based on option 2.

**Comment by yeyuqiang** [comment]
> > Why we need to struggle with option1(current code) here since we want to move forward to option 2.
> 
> No, we don't have to. But if we have some short cut ideas to make option 1 work. We can work on it. Otherwise, I am suggesting to refactor based on option 2.

My understand for option 1 is that the memory (100GB all) will be used as execution memory(set ratio as high as possible for execution memory.)
